core/modules/nat.cc

```diff
@@ -0,0 +1,239 @@
+  switch (ip->next_proto_id) {
+    case 0x06:
+      tcp->cksum = 0;
+      tcp->cksum = rte_ipv4_udptcp_cksum(ip, tcp);
+      break;
+    case 0x11:
+      udp->dgram_cksum = 0;
+      break;
+    case 0x01:
+      icmp->icmp_cksum = 0;
+      icmp->icmp_cksum = rte_ipv4_udptcp_cksum(ip, icmp);
+      break;
+    default:
+      VLOG(1) << "Unknown next_proto_id: " << ip->next_proto_id;
+  }
+  ip->hdr_checksum = rte_ipv4_cksum(ip);
+}
+
+inline static NAT::Flow parse_flow(struct ipv4_hdr *ip, void *l4) {
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(l4);
+  struct icmp_hdr *icmp = reinterpret_cast<struct icmp_hdr *>(l4);
+  NAT::Flow flow;
+
+  flow.proto = ip->next_proto_id;
+  flow.src_ip = ip->src_addr;
+  flow.dst_ip = ip->dst_addr;
+
+  switch (ip->next_proto_id) {
+    case 0x06:
+    case 0x11:
+      flow.src_port = udp->src_port;
+      flow.dst_port = udp->dst_port;
+      flow.icmp_ident = 0;
+      break;
+    case 0x01:
+      switch (icmp->icmp_type) {
+        case 0:
+        case 8:
+        case 13:
+        case 15:
+        case 16:
+          flow.src_port = 0;
+          flow.dst_port = 0;
+          flow.icmp_ident = icmp->icmp_ident;
+          break;
+        default:
+          VLOG(1) << "Unknown icmp_type: " << icmp->icmp_type;
+      }
+      break;
+    default:
+      VLOG(1) << "Unknown next_proto_id: " << ip->next_proto_id;
+  }
+  return flow;
+}
+
+inline static void stamp_flow(struct ipv4_hdr *ip, void *l4,
+                              const NAT::Flow &flow) {
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(l4);
+  struct icmp_hdr *icmp = reinterpret_cast<struct icmp_hdr *>(l4);
+
+  ip->src_addr = flow.src_ip;
+  ip->dst_addr = flow.dst_ip;
+
+  switch (flow.proto) {
+    case 0x06:
+    case 0x11:
+      udp->src_port = flow.src_port;
+      udp->dst_port = flow.dst_port;
+      break;
+    case 0x01:
+      switch (icmp->icmp_type) {
+        case 0:
+        case 8:
+        case 13:
+        case 15:
+        case 16:
+          icmp->icmp_ident = flow.icmp_ident;
+          break;
+        default:
+          VLOG(1) << "Unknown icmp_type: " << icmp->icmp_type;
+      }
+    default:
+      VLOG(1) << "Unknown next_proto_id: " << ip->next_proto_id;
+  }
+  compute_cksum(ip, l4);
+}
+
+void NAT::ProcessBatch(bess::PacketBatch *batch) {
+  gate_idx_t out_gates[bess::PacketBatch::kMaxBurst];
+  gate_idx_t incoming_gate = get_igate();
+
+  int cnt = batch->cnt();
+
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    // By default, pass packet
+    out_gates[i] = incoming_gate;
+
+    struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+    struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+    int ip_bytes = (ip->version_ihl & 0xf) << 2;
+    struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(
+        reinterpret_cast<uint8_t *>(ip) + ip_bytes);
+
+    // L4 protocol must be TCP, UDP, or ICMP
```

Comment:
It says TCP, UDP, or ICMP in the comment, but the code appears to only handle UDP since it assumes that it's a UDP header.  Do you plan to make it general for TCP/ICMP or am I misreading it?